rfq/negotiator.go

```diff
@@ -0,0 +1,424 @@
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/lightninglabs/taproot-assets/asset"
+	"github.com/lightninglabs/taproot-assets/fn"
+	"github.com/lightninglabs/taproot-assets/rfqmsg"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/routing/route"
+)
+
+// NegotiatorCfg holds the configuration for the negotiator.
+type NegotiatorCfg struct {
+	// PriceOracle is the price oracle that the negotiator will use to
+	// determine whether a quote is accepted or rejected.
+	PriceOracle PriceOracle
+
+	// OutgoingMessages is a channel which is populated with outgoing peer
+	// messages. These are messages which are destined to be sent to peers.
+	OutgoingMessages chan<- rfqmsg.OutgoingMsg
+
+	// ErrChan is a channel that is populated with errors by this subsystem.
+	ErrChan chan<- error
+}
+
+// Negotiator is a struct that handles the negotiation of quotes. It is a RFQ
+// subsystem. It determines whether a quote request is accepted or rejected.
+type Negotiator struct {
+	startOnce sync.Once
+	stopOnce  sync.Once
+
+	// cfg holds the configuration parameters for the negotiator.
+	cfg NegotiatorCfg
+
+	// assetSellOffers is a map (keyed on asset ID) that holds asset sell
+	// offers.
+	assetSellOffers map[asset.ID]SellOffer
+
+	// assetGroupSellOffers is a map (keyed on asset group key) that holds
+	// asset sell offers.
+	assetGroupSellOffers map[btcec.PublicKey]SellOffer
+
+	// assetSellOffersMtx is a mutex that protects the asset sell offers
+	// maps.
+	sellOffersMtx sync.RWMutex
+
+	// ContextGuard provides a wait group and main quit channel that can be
+	// used to create guarded contexts.
+	*fn.ContextGuard
+}
+
+// NewNegotiator creates a new quote negotiator.
+func NewNegotiator(cfg NegotiatorCfg) (*Negotiator, error) {
+	// If the price oracle is nil, then we will return an error.
+	if cfg.PriceOracle == nil {
+		return nil, fmt.Errorf("price oracle is nil")
+	}
+
+	return &Negotiator{
+		cfg: cfg,
+
+		assetSellOffers:      make(map[asset.ID]SellOffer),
+		assetGroupSellOffers: make(map[btcec.PublicKey]SellOffer),
+
+		ContextGuard: &fn.ContextGuard{
+			DefaultTimeout: DefaultTimeout,
+			Quit:           make(chan struct{}),
+		},
+	}, nil
+}
+
+// queryBidFromPriceOracle queries the price oracle for a bid price. It returns
+// an appropriate outgoing response message which should be sent to the peer.
+func (n *Negotiator) queryBidFromPriceOracle(peer route.Vertex,
+	assetId *asset.ID, assetGroupKey *btcec.PublicKey,
+	assetAmount uint64) (rfqmsg.OutgoingMsg, error) {
+
+	ctx, cancel := n.WithCtxQuitNoTimeout()
+	defer cancel()
+
+	oracleResponse, err := n.cfg.PriceOracle.QueryBidPrice(
+		ctx, assetId, assetGroupKey, assetAmount,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to query price oracle for "+
+			"bid: %w", err)
+	}
+
+	// If the bid price is nil, then we will return the error message
+	// supplied by the price oracle.
+	if oracleResponse.BidPrice == nil {
```

Comment:
This feels inverted; we're ignoring the `oracleResponse.Err`, unless the price is nil? Seems like we should always check that response err, and maybe modify our behavior if the price also happened to be unset.

Reply:
I'm going to refactor this function in the next PR. It will deal exclusively with price and wont return an RFQ message. This is because I'll use it when both buying and selling an asset.

I've added a TODO to remember to check the price oracle error correctly.